hal/src/thumbv7em/clock/v2/xosc.rs

```diff
@@ -422,9 +424,26 @@ where
         self.clock_switch = swben;
         self
     }
+
+    /// Modify hardware to realise the desired state
+    /// stored within the [`Xosc`]
+    ///
+    /// Returns the enabled Xosc
+    #[inline]
+    pub fn enable(mut self) -> Enabled<Xosc<X, M>, U0> {
+        self.token.reset();
```

Comment:
HW reset is a nice touch as it will harden this API against some misconfigured settings. Was lack of reset making an `Xosc` abstraction *very* unsound? I think you've mentioned that it did.

Reply:
I don't think we necessarily need a reset if we set *all* settings in the call to `.enable()`. But I don't think the existing implementation did that.

I think the only way to trigger a problem in this specific case would be to `.enable()` `CrystalMode`, `.disable()`, and then re-`.enable()` in `ClockMode`. I don't think anyone would ever actually do that. But it's something we should look out for.

Reply:
> I think the only way to trigger a problem in this specific case would be to .enable() CrystalMode, .disable(), and then re-.enable() in ClockMode. I don't think anyone would ever actually do that. But it's something we should look out for.

Why would it cause an issue? Source is changed and in clock mode, settings set in `CrystalMode::configure` are irrelevant (right?). I'm missing something probably. It's not like it matters but I'm just curious.

Reply:
Hmm, maybe it wouldn't have been an issue. I'm not sure. The previous implementation would set the crystal current, amplitude loop control and low buf gain *only* in crystal mode. If those settings have any effect on clock mode, they would not be reset when going from crystal mode to clock mode.

However, maybe those settings are completely ignored in clock mode. if that's the case, it probably doesn't matter. But I think I would still feel better knowing that we always start from a known state. Do you like that idea?

Reply:
This seems resolved?

I have no objection that we would like to remove possible unwanted statefulness and/or disconnects between abstraction model and hardware due to residual hardware configurations.